src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,111 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_SIZE = 10;
+    private T[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_SIZE];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elements.length) {
+            grow();
+        }
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index, must be 0 < i < size");
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        add(elements[size - 1]);
+        for (int i = size - 3; i >= index; i--) {
+            set(elements[i], i + 1);
+        }
+        set(value, index);
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        ArrayList<T> arrayList = (ArrayList<T>) list;
+        for (int i = 0; i < arrayList.size; i++) {
+            add(arrayList.elements[i]);
+        }
```

Comment:
```suggestion
        for (int i = 0; i < list.size(); i++) {
            add(list.get(i));
        }
```

we don't need to create `ArrayList<T>` here

Reply:
fixed